### PR TITLE
[SPARK-14914] Normalize Paths/URIs for windows. 

### PR DESCRIPTION
**Pull request by taoli91**
## What changes were proposed in this pull request?

The drive label and the back slash on windows' absolute path would confuse the URI scheme. This pull request mostly fix the path issue for windows. 
## How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by srowen** [comment]
@taoli91 please combine all your PRs into one and close the others.

**Comment by srowen** [comment]
Most of this PR is not the way to do it and adds lots of complexity. You're manually reimplementing a lot of logic in the JDK.

**Comment by taoli91** [comment]
@srowen Sure, I'll combine them into one. Could you elaborate more about what logic am I re-implementing?

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -1313,11 +1313,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    * supported for Hadoop-supported filesystems.
    */
   def addFile(path: String, recursive: Boolean): Unit = {
-    val uri = new URI(path)
-    val schemeCorrectedPath = uri.getScheme match {
-      case null | "local" => new File(path).getCanonicalFile.toURI.toString
-      case _ => path
-    }
+    val uri =Utils.resolveURI(path)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Maybe `val uri = Utils`

File: core/src/test/java/org/apache/spark/JavaAPISuite.java [hunk 2]
```diff
@@ -35,6 +35,7 @@
 import java.util.Set;
 import java.util.concurrent.*;
 
+import org.apache.spark.util.Utils;
```
Comment (HyukjinKwon) [inline core/src/test/java/org/apache/spark/JavaAPISuite.java]: Please see [databricks/scala-style-guide#imports](https://github.com/databricks/scala-style-guide#imports).

File: core/src/test/scala/org/apache/spark/metrics/InputOutputMetricsSuite.scala [hunk 3]
```diff
@@ -61,8 +61,7 @@ class InputOutputMetricsSuite extends SparkFunSuite with SharedSparkContext
     pw.close()
 
     // Path to tmpFile
-    tmpFilePath = "file://" + tmpFile.getAbsolutePath
-  }
+    tmpFilePath = Utils.resolveURI(tmpFile.getAbsolutePath).toString  }
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/metrics/InputOutputMetricsSuite.scala]: Maybe,
```scala
  tmpFilePath = Utils.resolveURI(tmpFile.getAbsolutePath).toString  
}
```
(See [databricks/scala-style-guide#imports](https://github.com/databricks/scala-style-guide#curly-braces)).

File: core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendSuite.scala [hunk 4]
```diff
@@ -25,20 +26,18 @@ import java.util.Collections
...
 import org.apache.mesos.Protos.Value.Scalar
 import org.mockito.{ArgumentCaptor, Matchers}
 import org.mockito.Matchers._
 import org.mockito.Mockito._
 import org.scalatest.mock.MockitoSugar
-
 import org.apache.spark.{LocalSparkContext, SparkConf, SparkContext, SparkFunSuite}
 import org.apache.spark.executor.MesosExecutorBackend
-import org.apache.spark.scheduler.{LiveListenerBus, SparkListenerExecutorAdded,
-  TaskDescription, TaskSchedulerImpl, WorkerOffer}
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackendSuite.scala]: Please do not remove newlines between imports, (See [databricks/scala-style-guide#imports] (https://github.com/databricks/scala-style-guide#imports))

File: core/src/test/scala/org/apache/spark/util/UtilsSuite.scala [hunk 5]
```diff
@@ -416,9 +416,9 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
     assertResolves("hdfs:/jar1,file:/jar2,jar3,jar4#jar5,path to/jar6",
       s"hdfs:/jar1,file:/jar2,file:$cwd/jar3,file:$cwd/jar4#jar5,file:$cwd/path%20to/jar6")
     if (Utils.isWindows) {
+      // It seems "#" is not valid for a local file path, and it will be escaped by the File.toURI
       assertResolves("""hdfs:/jar1,file:/jar2,jar3,C:\pi.py#py.pi,C:\path to\jar4""",
-        s"hdfs:/jar1,file:/jar2,file:$cwd/jar3,file:/C:/pi.py#py.pi,file:/C:/path%20to/jar4")
-    }
+        s"hdfs:/jar1,file:/jar2,file:$cwd/jar3,file:/C:/pi.py%23py.pi,file:/C:/path%20to/jar4")    }
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/util/UtilsSuite.scala]: Ditto.
```
    s"hdfs:/jar1,file:/jar2,file:$cwd/jar3,file:/C:/pi.py%23py.pi,file:/C:/path%20to/jar4")    
}
```

**Comment by srowen** [comment]
Broadly: you should be able to use the `File` API with forward-slash paths to describe and manipulate file locations, in a platform-independent way. Wherever you're changing Strings to Files is probably a good thing. Where you're manually handling paths with regexes seems suboptimal.

**Comment by taoli91** [comment]
@srowen

**Comment by taoli91** [comment]
> please combine all your PRs into one and close the others.
@srowen Double think about this problem, it seems the PR is too large if combining into one. I personally believe that it's better to categorize the changes and separate them.

**Comment by srowen** [comment]
The problem is that discussion gets split across many threads, when the changes have a logical relationship. I see that they're mostly distinct threads of change in this case. Some of these may end up not being merged or in reduced form, which makes them more combine-able. For now, leave them to see where they go

**Comment by HyukjinKwon** [comment]
(@taoli91 FYI, it would be great if you run `sbt scalastyle` or `./dev/run-tests` (this triggers style checking first) for style corrections before adding some commits. This will show up the corrections I said in the comments.)

**Comment by taoli91** [comment]
@HyukjinKwon Thanks for pointing out, I haven't noticed that there is an auto style check. Im going to have a vacation for a few days and I will fix them as soon as possible next week

**Comment by taoli91** [comment]
> Where you're manually handling paths with regexes seems suboptimal.
@srowen The regex parts are actually dealing with the URI. The default URI can't correctly handle the absolute Windows local path. It will recognize the drive label to be scheme. For example `new URI("D:\test")` will return a scheme of D. I can't find a JDK method to handle this problem, thus I have to use regex to find out the label and manually append a `/` in front of the path. E.g., convert `D:\test` to `/D:/test` before passing to the constructor of URI

**Comment by srowen** [comment]
Isn't the URI just `file:///D:/test`?

**Comment by taoli91** [comment]
@srowen `new FIle("D:\test").toURI.toString` can return `file://D/test` but `new URI("D:\test")` won't. But we should make sure that it is a local file before converting it into a file. Actually lots of method accepts a path string, and it will be converted to URI internally. Most test cases pass in path string instead of URI like string.

**Comment by srowen** [comment]
Yes, I'm assuming all of the inputs need to be parseable as a URI in order to keep this sensible. At least, that seems less problematic than trying to accept inputs like `D:\foo\bar`. `/foo/bar` is ambiguous since it could be a local or HDFS file, in which case it needs a scheme. This does probably mean some pieces of the code which deal with local files via string paths need some extra logic to parse as a URI and get the parsed path out of it. I think that's substantially what you're doing here, just wondering if we can make it simpler by making stronger input assumptions.

**Comment by vanzin** [comment]
Is this needed after #13868? pinging @avulanov

**Comment by avulanov** [comment]
@vanzin This PR partly addresses a broader problem that path construction is not consistent throughout Spark sources. Indeed, this is a potential source of bugs. Here is an example https://github.com/apache/spark/pull/13868#issuecomment-228932793. However, author addresses the problem with regex, which does not seem to be optimal as @srowen mentioned. I believe we should have a separate JIRA for making all path references and construction error-prone and consistent, and check everything including shell scripts. For example, https://github.com/apache/spark/blob/master/build/mvn#L161 contains a reference to `${MVN_BIN} -someParameters` which will fail if maven path contains a space.

**Comment by srowen** [comment]
Let's close this PR

**Comment by steveloughran** [comment]
If you are working with windows paths; Hadoop's Path class contains the code to do this, stabilised and addressing the corner cases

**Comment by steveloughran** [comment]
As #13868 does adopt `org.apache.hadoop.io.Path`, I don't see this patch being needed —though it may highlight some places where the new code may need applying

**Pull request closed without merging**